File: blob-module/src/BlobModule.ts
import { requireNativeModule } from "expo";
import { Blob } from "./BlobModule.types";

const NativeBlobModule: any = requireNativeModule("ExpoBlob");

export class ExpoBlob extends NativeBlobModule.Blob implements Blob {
	constructor(blob: string[], options?: BlobPropertyBag, other?: ExpoBlob) {
		super(blob, options, other)
	}
	
	// constructor(blob: string[]) {
	// 	super(blob)
	// }

	test(): string {
		return super.test();
	}

	opt(options : BlobPropertyBag) {
		return super.opt(options);
	}

	// constructor(blobParts?: any, options?: BlobPropertyBag) {
	// 	super(blobParts, options);
	// }

	// constructor(blob? : String | number | Blob[]) {
	// 	super(blob)
	// }

	// constructor() {
	// 	super()
	// }

	me(): Blob {
		return super.me()
	}

	slice(start?: number, end?: number, contentType?: string): Blob {
		return new ExpoBlob([], null, super.slice(start, end, contentType))
		return super.slice(start, end, contentType)
	}

	// either(sth : number | string): number | string {
	// 	return super.either(sth)
	// }

	// slice(start?: number, end?: number, contentType?: string): Blob {
	// 	const slicedBlob = super.slice(start, end, contentType);
	// 	const options: BlobPropertyBag = {
	// 		type: slicedBlob.type,
	// 		endings: slicedBlob.endings,
	// 	};
	// 	return new ExpoBlob(slicedBlob, options);
	// }

	async text(): Promise<string> {
		return Promise.resolve(super.text());
	}

	stream(): ReadableStream {
		const text = super.text();
		const encoder = new TextEncoder();
		const uint8 = encoder.encode(text);
		let offset = 0;
		return new ReadableStream<Uint8Array>({
			pull(controller) {
				if (offset < uint8.length) {
					controller.enqueue(uint8.subarray(offset));
					offset = uint8.length;
				} else {
					controller.close();
				}
			},
		});
	}
}

File: blob-module/android/src/main/java/expo/modules/blobmodule/BlobModule.kt
package expo.modules.blobmodule

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.types.Either
import expo.modules.kotlin.types.EitherOfThree

class BlobModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("ExpoBlob")

        Class(Blob::class) {
//            Constructor() { blob: List<String>, options: BlobOptionsBag?, other : Blob? ->
//                var type = BlobOptions().type
//                var endings = BlobOptions().endings
//
//                if (options != null) {
//                    type = options.type
//                    endings = options.endings
//                }
//
//                if (other == null) {
//                    Blob(blob, BlobOptions(type, endings))
//                } else {
//                    Blob(other.blob, other.options)
//                }
//            }

            Constructor() { blobParts: List<BlobPart>, options: BlobOptionsBag?->
                var type = BlobOptions().type
                var endings = BlobOptions().endings

                if (options != null) {
                    type = options.type
                    endings = options.endings
                }
                Blob(blobParts, BlobOptions(type, endings))
            }
            Property("size") { blob: Blob ->
                blob.size
            }

            AsyncFunction("text") { blob: Blob ->
                blob.text()
            }

            Property("type") { blob: Blob ->
                blob.options.type
            }

            Function("test") { blob: Blob ->
                "test"
            }
            Function("opt") { blob: Blob, options : BlobOptionsBag ->
                "${options.type}  ${options.endings}"
            }
            Function("slice") { blob: Blob, start: Int?, end: Int?, contentType: String? ->
                blob.slice(start ?: 0, end ?: blob.size, contentType ?: "")
            }
            Function("me") {blob:Blob -> blob}
        }
    }
}
